File: src/features/settings/export.ts
/* MISSION-071 — Export data layer. The user picks a format and a destination
   through the native save dialog (`@tauri-apps/plugin-dialog`), then the export
   runs as a background task: `export_media` resolves with the queued snapshot,
   `task-changed` events stream progress, and the typed `ExportReport`
   (`{ format, total, path }`) lands in the task's `result`. The task itself can
   be cancelled (dropping any partial file) via the shared task layer. */

import { useMutation } from "@tanstack/react-query";
import { save } from "@tauri-apps/plugin-dialog";
import { export_media, type TaskSnapshot } from "@/api";
import { useTask } from "@/features/tasks/api";

export type ExportFormat = "json" | "csv" | "markdown";

export interface ExportTarget {
  format: ExportFormat;
  path: string;
}

export const EXPORT_FORMATS: ExportFormat[] = ["json", "csv", "markdown"];

export function formatExtension(format: ExportFormat): string {
  return format === "markdown" ? "md" : format;
}

/** Spawn the export as a background task (MISSION-071). Resolves with the
    queued snapshot; follow progress through `useExportTask`. */
export function useExportMedia() {
  return useMutation({
    mutationFn: (target: ExportTarget): Promise<TaskSnapshot> =>
      export_media({ format: target.format, path: target.path }),
  });
}

/** Live snapshot of the export task (delegates to the shared `useTask`). */
export function useExportTask(taskId: string | null) {
  return useTask(taskId);
}

/** Native save dialog. Resolves with the chosen path or null when cancelled. */
export function pickExportPath(format: ExportFormat, filterName: string): Promise<string | null> {
  const extension = formatExtension(format);
  return save({
    title: "mylore",
    defaultPath: `mylore-library.${extension}`,
    filters: [{ name: filterName, extensions: [extension] }],
  });
}

File: src-tauri/src/commands/export.rs
//! Export commands (MISSION-071). Thin handler — the streaming write lives in
//! `application::export_service`. The export runs as a background task
//! (ARCHITECTURE §8): `export_media` opens the user-chosen path (from the
//! native save dialog on the frontend), spawns a `TaskKind::ExportFile` task
//! that streams the library row by row, and resolves with the initial snapshot.
//! Progress rides the `task-changed` events; cancellation drops the partial
//! file (the `PartialExport` guard removes it).

use sqlx::SqlitePool;
use tauri::command;
use tauri::State;
use tracing::info;

use std::path::Path;
use std::sync::Arc;

use crate::application::export_service::ExportService;
use crate::application::task_service::TaskManager;
use crate::domain::export::ExportFormat;
use crate::domain::task::{TaskError, TaskKind, TaskSnapshot};
use crate::error::AppError;

/// Export the whole library to `path` as a background task (MISSION-071).
/// `format` is `json` | `csv` | `markdown` (the save dialog's chosen
/// extension). Resolves with the queued snapshot; the `ExportReport` (format,
/// total, path) is the task's typed result on success.
#[command]
pub async fn export_media(
    tasks: State<'_, Arc<TaskManager>>,
    state: State<'_, SqlitePool>,
    format: String,
    path: String,
) -> Result<TaskSnapshot, AppError> {
    info!(format, path, "export_media invoked");
    let format = format.parse::<ExportFormat>()?;
    let pool = state.inner().clone();
    let title = format!("Export {} library", format.as_str());

    let id = tasks.spawn(TaskKind::ExportFile, title, move |reporter| async move {
        let service = ExportService::new(pool);
        reporter.progress(0, Some("Preparing export…".to_string()));

        let stream = service.stream_to_path(Path::new(&path), format, |done, total| {
            let percent = if total == 0 {
                100
            } else {
                done.checked_mul(100)
                    .and_then(|p| p.checked_div(total))
                    .unwrap_or(100) as u32
            };
            reporter.progress(percent, Some(format!("Exporting {done}/{total} titles")));
        });
        tokio::pin!(stream);

        tokio::select! {
            result = &mut stream => {
                let report = result.map_err(|error| TaskError::failed(error.to_string()))?;
                reporter.progress(100, Some("Export finished".to_string()));
                serde_json::to_value(&report)
                    .map_err(|error| TaskError::failed(error.to_string()))
            }
            _ = reporter.cancelled() => Err(TaskError::Cancelled),
        }
    });

    tasks
        .get(&id)
        .ok_or_else(|| AppError::internal("export task vanished"))
}
